test/unit/org/apache/cassandra/service/StorageServiceServerTest.java

```diff
@@ -122,9 +136,41 @@ public void testGetAllRangesEmpty()
     public void testSnapshot() throws IOException
     {
         // no need to insert extra data, even an "empty" database will have a little information in the system keyspace
+        StorageService.instance.clearSnapshot("snapshot");
         StorageService.instance.takeSnapshot("snapshot");
     }
 
+    @Test
+    public void testSnapshotJoiningNode() throws InterruptedException, IOException
+    {
+        StorageService ss = StorageService.instance;
+        TokenMetadata tmd = ss.getTokenMetadata();
+        tmd.clearUnsafe();
+        IPartitioner partitioner = RandomPartitioner.instance;
+        VersionedValue.VersionedValueFactory valueFactory = new VersionedValue.VersionedValueFactory(partitioner);
+
+        ArrayList<Token> endpointTokens = new ArrayList<Token>();
+        ArrayList<Token> keyTokens = new ArrayList<Token>();
+        List<InetAddress> hosts = new ArrayList<InetAddress>();
+        List<UUID> hostIds = new ArrayList<UUID>();
+
+        Util.createInitialRing(ss, partitioner, endpointTokens, keyTokens, hosts, hostIds, 7);
+
+        InetAddress bootstrappingNode = hosts.get(1);
+        Collection<Token> bootstrapTokens = Collections.singleton(keyTokens.get(3));
+        Gossiper.instance.injectApplicationState(bootstrappingNode, ApplicationState.TOKENS, valueFactory.tokens(bootstrapTokens));
+        ss.onChange(bootstrappingNode,
+                    ApplicationState.STATUS,
+                    valueFactory.bootstrapping(bootstrapTokens));
+
+        assertFalse(tmd.isMember(bootstrappingNode));
+        assertFalse(tmd.isLeaving(bootstrappingNode));
```

Comment:
what are we doing/checking from line 159 to here? I'm confused why we need this and not just the last 3 lines

Reply:
I was following the example of this test to alter the token metadata with a bootstrap instead of solely changing the operation mode: https://github.com/palantir/cassandra/blob/8a1cebc8c581d6935d7f00cee2b171259954c224/test/unit/org/apache/cassandra/service/LeaveAndBootstrapTest.java#L449

Do you think the test would be fine without this/only changing operation mode?

Reply:
If you try just keeping the last 3 lines, and reverting your changes in the code back to throwing an error if the node is bootstrapping, does this test fail? If so I think we can remove this stuff. If it passes then I guess we need this stuff to get to that part of the code. Does that make sense?